packages/hardhat-kadena/contracts/Redeem.yul

```diff
@@ -0,0 +1,177 @@
+    let paramsSize := 0xe3
+
+    if iszero(eq(calldatasize(), paramsSize)) {
+      error(0x01)
+    }
+
+    // calldataload + shift seems to incur the lowest gas cost
+    // (note that shr is cheaper than div)
+    let amount := calldataload(0x0)
+    let version := shr(0xf0, calldataload(0x20))
+    let targetChain := shr(0xe0, calldataload(0x22))
+    let targetAccount := shr(0x60, calldataload(0x26))
+    let xChanBalance := calldataload(0x3a)
+    let xChanId := calldataload(0x5a)
+    let rootHash := calldataload(0x7a)
+    let sigV := shr(0xf8, calldataload(0x9a))
+    let sigR := calldataload(0x9b)
+    let sigS := calldataload(0xbb)
+    let rootTime := shr(0xc0, calldataload(0xdb))
+
+    /* ********************************************************************** */
+    /* Check Version */
+
+    if iszero(eq(version, 0x0)) {
+      error(0x02)
+    }
+
+    /* ********************************************************************** */
+    /* Check Target ChainId */
+
+    let chainwebChainIdContract := 0x9b02c3e2df42533e0fd166798b5a616f59dbd2cc
+    if iszero(staticcall(gas(), chainwebChainIdContract, 0, 0, add(chainIdPtr, 0x1c), 0x4)) {
+      error(0x03)
+    }
+    if iszero(eq(mload(chainIdPtr), targetChain)) {
+      error(0x04)
+    }
+
+    /* ********************************************************************** */
+    /* Compute Message Hash */
+
+    calldatacopy(msgPtr, msgOffset, msgSize)
+    mstore(msgHashPtr, keccak256(msgPtr, msgSize))
+
+    /* ********************************************************************** */
+    /* Check Signature */
+
+    // copy v value (right aligned)
+    calldatacopy(add(sigPtr, 0x1f), sigOffset, 0x01)
+    // copy r and s value
+    calldatacopy(add(sigPtr, 0x20), add(sigOffset, 0x01), 0x40)
+
+    let ecrecoverContract := 0x01
+    if iszero(staticcall(gas(), ecrecoverContract, msgHashPtr, 0x80, sigAddrPtr, 0x20)) {
+      error(0x05)
+    }
+    if iszero(eq(mload(sigAddrPtr), REDEEM_KEY_ADDR)) {
+      error(0x06)
+    }
+
+    /* ********************************************************************** */
+    /* Query Header Oracle */
+
+    // Use EIP-4788 to check whether a header is included on chain.  For that
+    // the timestamp of the header is needed. This information is merly
+    // informational and does not need to be verfied.
+    //
+    // Also note that only the most recent 8191 headers are available which is
+    // about 3 days on the Kadena mainnet.
+
+    let beaconRootsContract := 0x000F3df6D732807Ef1319fB7B8bB8522d0Beac02
+    mstore(rootTimePtr, rootTime)
+    if iszero(staticcall(gas(), beaconRootsContract, rootTimePtr, 0x20, rootPtr, 0x20)) {
+      error(0x07)
+    }
+    if iszero(eq(mload(rootPtr), rootHash)) {
+      error(0x08)
+    }
+
+    /* ********************************************************************** */
+    /* Check Amount */
+
+    let redeemedAmount := sload(xChanId)
+    let maxAmount := sub(xChanBalance, redeemedAmount)
```

Comment:
Add comment why the use of `sub` here is safe without checking for an arithmetic underflow.